The start of documentation 2.0 (#3673)
CliMA/Oceananigans.jl: docs/src/grids.md

The "grid" encodes fundamental information: the shape of the domain we're simulating in, the way that domain is divided into a mesh of finite volumes, the machine architecture (CPU, GPU, lots of GPUs), and the precision of the numbers we would like to use to represent ocean-flavored fluid variables (double precision or single precision).

One of the simplest grids we can make divides a three-dimensional rectangular domain -- or "a box" --- into evenly-spaced cells.
To create such a grid on the CPU (the machine that we typically run julia on, basically), we write

```jldoctest grids
architecture = CPU()

grid = RectilinearGrid(architecture,
                       topology = (Periodic, Periodic, Bounded),
                       size = (16, 8, 4),
                       x = (0, 64),
                       y = (0, 32),
                       z = (0, 8))

# output
16×8×4 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── Periodic x ∈ [0.0, 64.0) regularly spaced with Δx=4.0
├── Periodic y ∈ [0.0, 32.0) regularly spaced with Δy=4.0
└── Bounded  z ∈ [0.0, 8.0]  regularly spaced with Δz=2.0
```

> This simple grid

Reviewer: ```suggestion
This simple grid:
```

Author: "this simple grid: uses the CPU" isn't correct right?

Author: see here:

https://style.mla.org/vertical-lists-in-mla-style/

under "Lists That Continue the Sentence Introducing Them"

Reviewer: great! even if it was merely your preference I'd be convinced!

> * Uses the CPU. To make a grid on the GPU --- which means that computations on the grid will be conducted using a GPU connected to our CPU, if one is available --- we write `architecture = GPU()`.

Author: I like it

* Has a domain that's "periodic" in ``x, y``, but bounded in ``z``. More on what that means, exactly, in a bit.
* Has `16` cells in `x`, `8` cells in `y`, and `4` cells in `z`. That means there are ``16 \times 8 \times 4 = 512`` cells in all.

> * Has an `x` dimension that spans from `x=0`, to `x=64`. And `y` spans `y=0` to `y=32`, and `z` spans `z=0` to `z=8`.

Reviewer: ```suggestion
* Has an ``x`` dimension that spans from `x=0`, to `x=64`, ``y`` spans from `y=0` to `y=32`, and ``z`` spans `z=0` to `z=8`.
```

Author: I'm referring to the kwargs here

Reviewer: oh OK!

* Has cells that are all the same size, dividing the ``16 \times 8 \times 4`` box into ``4 \times 4 \times 2`` cells. Note that length units are whatever is used to construct the grid, so it's up to the user to make sure that all inputs use consistent units.

Next we illustrate the construction of a grid on the _GPU_ (for this to work, a GPU has
to be available and `CUDA.jl` must be working --- for more about that, see the
[`CUDA.jl` documentation](https://cuda.juliagpu.org/stable/)).
The grid is two-dimensional in ``x, z`` (it's missing the `y`-dimension), and has variably-spaced
cell interfaces in the `z`-direction,

```@setup grids_gpu
using Oceananigans
```

```jldoctest grids_gpu
architecture = GPU()
z_faces = [0, 1, 3, 6, 10]

grid = RectilinearGrid(architecture,
                       topology = (Periodic, Flat, Bounded),
                       size = (10, 4),
                       x = (0, 20),
                       z = z_faces)

# output
10×1×4 RectilinearGrid{Float64, Periodic, Flat, Bounded} on GPU with 3×0×3 halo
├── Periodic x ∈ [0.0, 20.0)      regularly spaced with Δx=2.0
├── Flat y
└── Bounded  z ∈ [0.0, 10.0]      variably spaced with min(Δz)=1.0, max(Δz)=4.0

> ```

Reviewer: so this will fail if docs built don't run on GPUs

Author: I changed it so the docs build on the GPU

> The `y`-dimension is marked `Flat` by writing `topology = (Periodic, Flat, Bounded)`.
> This means that nothing varies in `y`; derivatives in `y` are 0.

Reviewer: ```suggestion
For the above grid, the ``y``-dimension is marked `Flat` by specifying `topology = (Periodic, Flat, Bounded)`.
This means that nothing varies in ``y``; derivatives in ``y`` are 0.
```

> This also means that the kwarg that specifies the `y`-domains may be omitted, and that

Reviewer: ```suggestion
This also means that the keyword argument (or short, kwarg) that specifies the ``y``-domains may be omitted, and that
```

the `size` has only two elements rather than 3 as in the first example.
Notice that in the stretched cell interfaces specified by `z_interfaces`, the number of
vertical cell interfaces is `Nz + 1 = length(z_interfaces) = 5`, where `Nz = 4` is the number
of cells in the vertical.

## The nuts and bolts of building a grid

> These examples illustrate how the definition of any grid requires

Reviewer: ```suggestion
These examples illustrate how the definition of any grid requires:
```

Author: confused whether this is correct

Reviewer: ok, perhaps it's a style choice -- I'll leave it up to you to decide!

Author: I think a colon is incorrect here!

Author: (because the bullet points "finish" the sentence started above)

Reviewer: gotta

    - [`Periodic`](@ref), which means that the dimension circles back onto itself: information leaving the left side of the domain re-enters on the right.
    - [`Bounded`](@ref), which means that the two sides of the dimension are either impenetrable (solid walls), or "open", representing a specified external state.
    - [`Flat`](@ref), which means nothing can vary in that dimension, reducing the overall dimensionality of the grid.
* Defining the number of cells that divide each dimension. The number of cells, with or without explicit specification of the cell interfaces, determines the spatial resolution of the grid.
* The representation of floating point numbers, which can be single-precision (`Float32`) or double precision (`Float64`).

Let's dive into each of these options in more detail.

### Specifying the machine architecture

The positional argument `CPU()` or `GPU()`, specifies the "architecture" of the simulation.

> By writing `architecture = GPU()`, any fields constructed on `grid` will store their data on
> an Nvidia [`GPU`](@ref), if one is available. By default, the grid will be constructed on

Reviewer: ```suggestion
By using `architecture = GPU()`, any fields constructed on `grid` store their data on
an Nvidia [`GPU`](@ref), if one is available. By default, the grid will be constructed on
```

                       x = (0, 64),
                       y = (0, 32),
                       z = (0, 8))

# output
16×8×4 RectilinearGrid{Float32, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── Periodic x ∈ [0.0, 64.0) regularly spaced with Δx=4.0
├── Periodic y ∈ [0.0, 32.0) regularly spaced with Δy=4.0
└── Bounded  z ∈ [0.0, 8.0]  regularly spaced with Δz=2.0
```

Single precision should be used with care.
Users interested in performing single-precision simulations should get in touch via
[Discussions](https://github.com/CliMA/Oceananigans.jl/discussions),
and should subject their work to extensive testing and validation.

For more examples see [`RectilinearGrid`](@ref) and [`LatitudeLongitudeGrid`](@ref).

## Distributed grids

To run the next examples, we have to launch julia with MPI and things are a bit more involved.
For best results, create [a new environment](https://pkgdocs.julialang.org/v1/environments/)
in an empty folder, and then add both `Oceananigans.jl` and `MPI.jl` to the new environment.
Pasting this snipped into the terminal should do the trick:

```bash
mkdir OceananigansDistributedExamples
cd OceananigansDistributedExamples

> touch Project.toml
> julia --project -e 'using Pkg; Pkg.add(["Oceananigans", "MPI"])'

Reviewer: `touch` is unix command; this will work in more machines:

```suggestion
julia --project -e 'using Pkg; using Pkg; Pkg.activate("."); Pkg.add(["Oceananigans", "MPI"])'
```

Author: I think not

                       z = z_faces)

# output
ERROR: ArgumentError: Cannot create a GPU grid. No CUDA-enabled GPU was detected!
Stacktrace:
 [1] RectilinearGrid(architecture::GPU, FT::DataType; size::Tuple{Int64, Int64}, x::Tuple{Int64, Int64}, y::Nothing, z::Vector{Int64}, halo::Nothing, extent::Nothing, topology::Tuple{DataType, DataType, DataType})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:266
 [2] RectilinearGrid
   @ ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:255 [inlined]
 [3] top-level scope

>    @ none:1

Reviewer: what's this?
